packages/product/src/services/product-module-service.ts

```diff
@@ -405,6 +398,229 @@ export default class ProductModuleService<
+          string,
+          ProductTypes.CreateProductOptionDTO[]
+        >()
+
+        const productsData = await Promise.all(
+          data.map(async (product) => {
+            const productData = { ...product }
+            const { variants, options } = productData
+            delete productData.options
+            delete productData.variants
+
+            if (!isDefined(productData.id)) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `Cannot update product without id`
+              )
+            }
+
+            productVariantsMap.set(productData.id, variants ?? [])
+            productOptionsMap.set(productData.id, options ?? [])
+
+            if (productData.is_giftcard) {
+              productData.discountable = false
+            }
+
+            await this.setupProductDataForImages(productData, sharedContext)
+            await this.upsertProductTags(productData, sharedContext)
+            await this.upsertProductTypes(productData, sharedContext)
+
+            return productData as ProductServiceTypes.UpdateProductDTO
+          })
+        )
+
+        const products = await this.productService_.update(
+          productsData,
+          sharedContext
+        )
+
+        const productByIdMap = new Map<string, TProduct>(
+          products.map((product) => [product.id, product])
+        )
+
+        const productOptionsData = [...productOptionsMap]
+          .map(([id, options]) => options.map((option) => ({
+            ...option,
+            product: productByIdMap.get(id)!,
+          })))
+          .flat()
+
+        const productOptions = await this.productOptionService_.create(
+          productOptionsData,
+          sharedContext
+        )
+
+        let productVariantIdsToDelete: string[] = []
+        const productVariantsToCreateMap = new Map<
+          string,
+          ProductTypes.CreateProductVariantDTO[]
+        >()
+
+        const productVariantsToUpdateMap = new Map<
+          string,
+          ProductTypes.UpdateProductVariantDTO[]
+        >()
+
+        for (const [productId, variants] of productVariantsMap) {
+          const variantsToCreate: ProductTypes.CreateProductVariantDTO[] = []
+          const variantsToUpdate: ProductTypes.UpdateProductVariantDTO[] = []
+          const existingVariants = existingProductVariantsMap.get(productId)
+
+          variants.forEach((variant) => {
+            const isVariantIdDefined = ("id" in variant) && isDefined(variant.id)
+            const existingProductVariant = existingVariants
+              ?.find((existingVariant) => isVariantIdDefined && (existingVariant.id === variant.id))
+
+            if (isVariantIdDefined && !existingProductVariant) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `ProductVariant "${variant.id}" not found for product "${productId}"`
+              )
+            }
```

Comment:
todo: rm based on the comment in the initialisation of the map which now would throw earlier preventing us from doing plenty of things before knowing something is wrong

Reply:
removed